### PR TITLE
fix: Reentrancy guards and project-to-terminal checks

### PR DESCRIPTION
**Pull request by dhyon**
- added reentrancy guards to terminal pay and addToBalance methods to address audit issue PVE005
- added project-to-terminal checks in terminal pay and addToBalance methods to prevent situations where callers can mistakenly call the wrong terminal + tests

### CODE REVIEW

File: contracts/JBETHPaymentTerminal.sol [hunk 1]
```diff
@@ -246,7 +247,12 @@ contract JBETHPaymentTerminal is
     bool _preferClaimedTokens,
     string calldata _memo,
     bytes calldata _delegateMetadata
-  ) external payable override {
+  ) external payable override nonReentrant {
+    // Verify that this terminal is a terminal of provided project ID
+    if (!directory.isTerminalOf(_projectId, this)) {
```
Comment (mejango) [inline contracts/JBETHPaymentTerminal.sol:252]: you can abstract this to an `requireTerminalMatch` modifier and reuse in addToBalance, whatchu think?

Comment (dhyon) [reply contracts/JBETHPaymentTerminal.sol:252]: sure that's DRYer

**Comment by drgorillamd** [comment]
noice!

**Review by mejango — APPROVED** [review]
awesome

**Pull request merged**